Add SORT_STRICT option to array_unique()

cscott opened the pull request:
Provide a way to request the use of strict comparison (===) when using `array_unique`.

nikic commented on ext/standard/array.c:
```diff
@@ -349,6 +350,22 @@ static zend_always_inline int php_array_data_compare_unstable_i(Bucket *f, Bucke
+	if (c != 0) {
+		return c;
+	}
+	// types must be different; compare based on type.
```
Arrays and objects have different strict/weak comparison semantics. You can have two arrays/objects that are `!==` and `==`, and you need to specify a well-defined ordering for these.

cscott replied:
For arrays these are two arrays which have the exact same contents.  These have the same semantics for `==` and `===`:
```
1] boris> $a = [1,2,3];
 → array(
  0 => 1,
  1 => 2,
  2 => 3
)
[2] boris> $b = [1,2,3];
 → array(
  0 => 1,
  1 => 2,
  2 => 3
)
[3] boris> $a == $b;
 → true
[4] boris> $a === $b;
 → true
```
So that shouldn't be a problem.  The object case /is/ an issue:
```
[6] boris> $a = (object)$a;
 → object(stdClass)(
  0 => 1,
  1 => 2,
  2 => 3
)
[7] boris> $b = (object)$b;
 → object(stdClass)(
  0 => 1,
  1 => 2,
  2 => 3
)
[8] boris> $a == $b;
 → true
[9] boris> $a === $b;
 → false
```
I propose to compare these based on `spl_object_id`. That should work, right?

cscott commented:
Ok, latest push should break ties in all situations where the values are `==` but not `===`.

nikic commented on ext/standard/array.c:
```diff
@@ -349,6 +350,65 @@ static zend_always_inline int php_array_data_compare_unstable_i(Bucket *f, Bucke
+
+	default:
+		ZEND_ASSERT(0 && "Value types should have consistent == and ===");
+		return 0;
```
This still doesn't cover everything, for example consider `"0"` and `"0.0"`. I would recommend to explicitly write out the comparison function and not rely on zend_compare at all.

cscott replied:
Ok, done.  Good advice.

nikic commented on ext/standard/array.c:
```diff
@@ -349,6 +350,102 @@ static zend_always_inline int php_array_data_compare_unstable_i(Bucket *f, Bucke
+
+	case IS_OBJECT:
+	{
+		// Start with a recursive comparison like for arrays, for consistency.
```
Why doesn't this only compare the object ID? That should be sufficient to establish an ordering.

cscott replied:
Because the object ID leads to nondeterministic behavior.  In my experience, having an array be sorted slightly differently every time you run the program is the sort of thing that leads to headaches.  Starting with a recursive comparison of attributes and properties is consistent with the way arrays are sorted, and the way that objects and arrays are compared with == (ie, that's the behavior of `zend_std_compare_objects`).  Comparing class names is an extra attempt to avoid non-determinism.  The object ID is only used as a fallback if your objects really are identical in all other ways.  Ie, the way I've implemented it here the order of the `==` elements will be nondeterministic, but all the other `!=` elements will have a consistent order from run to run.

cscott replied:
(in latest patch added test cases to document predictable order of non-identical objects, dropped the class name comparison.)

cscott replied:
(and to be clear by "nondeterminism" I mean that adding a new object allocation somewhere in your program, or reordering some code which performs allocations, can cause a nonlocal change in the sort order elsewhere.  I believe in PHP the object IDs are actually, strictly-speaking, deterministic given identical inputs to a program, but in practice in large programs it "feels" nondeterministic because of the effect of nonlocal changes.)

nikic replied:
Depends on what the goal here is. I don't think SORT_STRICT is particularly useful outside array_unique(). And array_unique() just needs some arbitrary order, so it should be the most efficient such order.

It may be worthwhile to accept SORT_STRICT in array_unique() only, to avoid exposing any particular strict ordering to userland -- it becomes an unobservable implementation detail.

cscott replied:
Yes, I started out thinking about just `array_unique`, but once you define it as a constant it seems to invite being used for all the different array methods which accept flags.  And I can think of use cases where a strict ordering is useful, for example anywhere where string and numeric values are mixed but you don't want the `"0"`, `"0.0"`, and `0` keys to all be considered the same. (`SORT_STRING` will conflate `"0"` and `0`, and `SORT_DEFAULT` and `SORT_NUMERIC` will conflate `"0"` and `"0.0"`).  Similarly if you want `null` to be a valid value sorted distinct from `""`.
"Sort, but don't do *any* conversion on the keys" actually seems like a genuinely-useful feature.
```
$a = [ "0", "0.0", "", null, "", "0.0", "0" ];
$flags = 0; // Nothing we can put here will put all identical values together
asort($a, $flags);
var_dump($a);
```

I'm not convinced that the sort order for `array_unique` is an "unobservable implementation detail" -- I think the actual sort order chosen is still revealed by the order of the keys in the returned array.  I have to think about that some more.

nikic commented on ext/standard/array.c:
```diff
@@ -349,6 +350,102 @@ static zend_always_inline int php_array_data_compare_unstable_i(Bucket *f, Bucke
+	ZVAL_DEREF(z2);
+	// The most important thing about this comparison mode is that the result
+	// is 0 when zend_is_identical, and non-zero otherwise.
+	if (zend_is_identical(z1, z2)) {
```
As you're enumerating all cases below anyway, I'd drop this check and handle equality below as well. For arrays that would save one deep comparison.

cscott replied:
Yes, but it makes it harder to maintain the invariant, and less obvious if there's some subtle bug somewhere that causes the output of zend_is_identical and this function to diverge.  It's easier (IMO) to determine by inspection that the only way a zero can be returned from this function is if zend_is_identical returns zero.  If you're worried about performance I could probably exclude objects and arrays from this comparison -- arrays involve a call to zend_hash_compare anyway so the "obvious by inspection it can't return 0" thing isn't as obviously true there anyway.

cscott replied:
(avoided the deep comparison in latest patch, as requested.)

cscott commented:
Updated the patch to simplify and correct the comparison: we no longer use class name for objects; we avoid the redundant deep inspection for arrays; and we use ordered comparison of properties for objects.  Added additional test cases to document the intended behavior, including the deterministic ordering of objects with non-identical properties.

TysonAndre reviewed (commented):
Coincidentally, I'd also been working on stable sorting lately and didn't remember your PR, found it again when searching past bugs/proposals.

https://github.com/TysonAndre/pecl-teds/blob/0.5.1/teds.c#L567-L657 had some differences in the approach

1. I'd agree with earlier comments. If this is only meant to be for array_unique, then it should be rejected for other sort functions, and possibly given a different name (ARRAY_FILTER_STRICT, ARRAY_FILTER_SORT_STRICT_ORDER, or something)

2. This should handle NAN. Operators involving NAN are always false in C/PHP. This would break the ordering of floats in arrays. (except `NAN !== NAN` is true)

E.g. if `2.0 < NAN` is false, and `NAN < 2.0` is false, we may get the ordering 2.0, NAN, 2.0 and have duplicate values of 2.

```
php > var_export(1.0 <=> NAN);
1
php > var_export(NAN <=> NAN);
1
php > var_export(NAN <=> 1.0);
1
```

3. Some object types make no sense to compare by property set, e.g. Closure or GMP have no properties to compare.
```
php > $x = gmp_init(1);
php > var_export((string)$x);
'1'
php > var_export(get_object_vars($x));
array (
)
php > var_export($x);
GMP::__set_state(array(
))
```

Maybe changing the second argument of `array_unique` to `int|Closure` would help (`Closure(int $a, int $b): int` for comparator, like usort), so that users could specify custom comparators for objects.
- Not sure if there'd be enough interest in allowing this, but it seems to have use cases, e.g. for `array_unique($items, fn($a, $b) => $a->id <=> $b->id);` without creating an extra array of just ids and intersecting keys
- (this may need extra checks to guard against side effects of functions during sorting and stop/recover if there's an exception?)

EDIT: This was apparently proposed in https://externals.io/message/70060 but not really followed up on. 

4. If this is exposed for `*sort` to end users, then `1.5 < (int)2` should be true (3/2 < 4/2 would be intuitive), and compare(2, 2.0) != 0 should also be true. See comment below `t1 != t2`

TysonAndre commented:
Definitely not suggesting switching approach or adding to scope, just making a note of an idea I've had for how php could expose object comparisons safely in arrays of objects - Implementing an extensible Comparator class seems doable but would require much more familiarity with internals to guard against comparison callbacks mutating parent objects/arrays while recursing on zvals, and possibly detecting infinite recursion (adding and decrementing reference counts with GC_ADDREF/GC_DELREF)

- A `Comparator`/`Collections\Comparator` may be of use for implementing things like a extensible SortedSet in userland or internally, to handle arrays of objects and other values.


Related past/current RFCs tried to change objects themselves:

- https://wiki.php.net/rfc/comparator_interface https://wiki.php.net/rfc/comparable proposed a Comparable interface, Comparator seems like the wrong name for an interface and were abandoned.
- https://wiki.php.net/rfc/user_defined_operator_overloads#vote is open but currently does not meet the required 2/3 voting threshold. Also see https://wiki.php.net/rfc/userspace_operator_overloading 

```php
// Extensible internal class - user-defined classes can override compare, compareObjects, compareStrings
// $c = Closure::fromCallable(new ComparatorOrSubclass()); can be used to create closures to pass elsewhere, e.g. array_unique($values, $c);
// Allows tracking state and customizations while comparing.
class Comparator {
    /** Compares two top-level values but not inner values, in an overridable way (e.g. to allow ReverseComparator, custom behavior only for top-level objects, etc.) */
    public function __invoke(mixed $a, mixed $b): int {
    }
    /** Compares two objects. Can be extended to customize for specific use cases. Defaults to .... */
    public function compareObjects(object $a, object $b): int {
    }
    /** Compares two strings. Defaults to same ordering as strcmp. */
    public function compareStrings(string $a, string $b): int {
    }
    // I expect php to remove resources at some point.
}
```
